src/op_handlers/ret.rs

```diff
@@ -25,6 +25,9 @@ fn get_result(
     }
     let register = vm.get_register(reg_index);
     let result = get_forward_memory_pointer(register.value, vm, register.is_pointer)?;
+    if !vm.current_context()?.is_kernel() && result.page == vm.current_context()?.calldata_heap_id {
+        return Err(EraVmError::NonValidForwardedMemory);
```

Comment:
Are you sure you have to return an err here? I think in this case you should just return a new `TaggedValue::new_pointer(U256::zero())`

Reply:
On the vm2 they do a panic 
```
let result = get_far_call_calldata(raw_abi, is_pointer, vm, false).filter(|pointer| {
                vm.state.current_frame.is_kernel
                    || pointer.memory_page != vm.state.current_frame.calldata_heap
            });

            if result.is_none() {
                return_type = ReturnType::Panic;
            }
```
Doing this will call the `inexplicit_panic` that performs the correct behaviour